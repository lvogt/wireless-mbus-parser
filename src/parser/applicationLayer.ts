import {
  AES_BLOCK_SIZE,
  CI_RESP_0,
  CI_RESP_4,
  CI_RESP_12,
  CI_RESP_COMPACT,
  CI_RESP_PRIOS,
  CI_RESP_SML_4,
  CI_RESP_SML_12,
  CI_RESP_TECHEM,
} from "@/helper/constants";
import { calcKenc, checkAflMac, decryptInPlace } from "@/helper/crypto";
import {
  decodeManufacturer,
  getDeviceState,
  getDeviceType,
  getMeterId,
  isLinkLayer,
} from "@/helper/helper";
import { decodePriosApplicationLayer } from "@/parser/priosApplicationLayer";
import { decodeTechemApplicationLayer } from "@/parser/techemApplicationLayer";
import type {
  ApplicationLayer,
  ApplicationLayer4,
  ApplicationLayer12,
  ApplicationLayerCompact,
  AuthenticationAndFragmentationLayer,
  Config,
  ConfigMode5,
  ConfigMode7,
  ConfigMode13,
  LinkLayer,
  ParserState,
  WiredLinkLayer,
} from "@/types";

function getApplicationLayerCompact(
  data: Buffer,
  pos: number,
  aplOffset: number
): {
  newPos: number;
  apl: ApplicationLayerCompact;
} {
  const headerCrc = data.readUInt16LE(pos);
  pos += 2;
  const frameCrc = data.readUInt16LE(pos);
  pos += 2;

  return {
    newPos: pos,
    apl: {
      ci: CI_RESP_COMPACT,
      offset: aplOffset,
      headerCrc: headerCrc,
      frameCrc: frameCrc,
    },
  };
}

function getApplicationLayer4(
  data: Buffer,
  pos: number,
  aplOffset: number
): {
  newPos: number;
  apl: ApplicationLayer4;
} {
  const accessNo = data[pos++];
  const statusCode = data[pos++];
  const status = getDeviceState(statusCode);

  const { config, newPos } = getConfig(data, pos);
  return {
    newPos,
    apl: {
      ci: CI_RESP_4,
      offset: aplOffset,
      accessNo,
      statusCode,
      status,
      config,
    },
  };
}

function getApplicationLayer12(
  data: Buffer,
  pos: number,
  aplOffset: number
): {
  newPos: number;
  apl: ApplicationLayer12;
} {
  const meterId = data.readUInt32LE(pos);
  pos += 4;
  const meterManufacturer = data.readUInt16LE(pos);
  pos += 2;
  const meterVersion = data[pos++];
  const meterDevice = data[pos++];

  const apl4 = getApplicationLayer4(data, pos, aplOffset);

  return {
    newPos: apl4.newPos,
    apl: {
      ...apl4.apl,
      ci: CI_RESP_12,
      meterId,
      meterIdString: getMeterId(meterId),
      meterManufacturer,
      meterManufacturerString: decodeManufacturer(meterManufacturer),
      meterVersion,
      meterDevice,
      meterDeviceString: getDeviceType(meterDevice),
    },
  };
}

function getDummyLinkLayer(ll: LinkLayer | WiredLinkLayer): LinkLayer {
  if (isLinkLayer(ll)) {
    return ll;
  } else {
    return {
      lField: ll.lField,
      mField: 0,
      cField: ll.cField,
      aField: ll.aField,
      version: 0,
      type: 0,

      addressRaw: Buffer.alloc(8),
      aFieldRaw: Buffer.alloc(6),

      manufacturer: "AAA",
      typeString: "unknown",
      meterId: "00000000",
    };
  }
}

function mockLinkLayerFromApplicationLayer(
  apl: ApplicationLayer,
  ll: WiredLinkLayer | LinkLayer
): LinkLayer {
  if (isLinkLayer(ll)) {
    return ll;
  }

  if (apl.ci !== CI_RESP_12) {
    return getDummyLinkLayer(ll);
  }

  // copy over application data meter and address info to link layer data for compatibility
  const addressRaw = Buffer.alloc(8);
  addressRaw.writeUInt32LE(apl.meterManufacturer, 0);
  addressRaw.writeUInt32LE(apl.meterId, 2);
  addressRaw[6] = apl.meterVersion;
  addressRaw[7] = apl.meterDevice;

  return {
    lField: ll.lField,
    mField: apl.meterManufacturer,
    cField: ll.cField,
    aField: ll.aField,
    version: apl.meterVersion,
    type: apl.meterDevice,

    addressRaw: addressRaw,
    aFieldRaw: addressRaw.subarray(2),

    manufacturer: apl.meterManufacturerString,
    typeString: apl.meterDeviceString,
    meterId: apl.meterIdString,
  };
}

function getConfig5(mode: 0 | 5, cw: number): ConfigMode5 {
  return {
    mode: mode,
    bidirectional: (cw & 0b1000000000000000) !== 0,
    accessability: (cw & 0b0100000000000000) !== 0,
    synchronous: (cw & 0b0010000000000000) !== 0,
    encryptedBlocks: (cw & 0b0000000011110000) >> 4,
    content: (cw & 0b0000000000001100) >> 2,
    hopCounter: cw & 0b0000000000000011,
  };
}

function getConfig7(mode: 7, cw: number, cwe: number): ConfigMode7 {
  return {
    mode: mode,
    content: (cw & 0b1100000000000000) >> 14,
    encryptedBlocks: (cw & 0b0000000011110000) >> 4,
    kdfSel: (cwe & 0b00110000) >> 4,
    keyid: cwe & 0b00001111,
  };
}

function getConfig13(mode: 13, cw: number, cwe: number): ConfigMode13 {
  return {
    mode: mode,
    content: (cw & 0b1100000000000000) >> 14,
    encryptedBytes: cw & 0b0000000011111111,
    protoType: cwe & 0b00001111,
  };
}

function getConfig(
  data: Buffer,
  pos: number
): {
  config: Config;
  newPos: number;
} {
  const cw = data.readUInt16LE(pos);
  pos += 2;

  const mode = (cw & 0b0001111100000000) >> 8;
  switch (mode) {
    case 0:
    case 5:
      return { config: getConfig5(mode, cw), newPos: pos };
    case 7:
      return { config: getConfig7(mode, cw, data[pos]), newPos: pos + 1 };
    case 13:
      return { config: getConfig13(mode, cw, data[pos]), newPos: pos + 1 };
    default:
      throw new Error(`Unhandled config mode ${mode}`);
  }
}

function createIvMode7() {
  return Buffer.alloc(16, 0x00);
}

function createIvMode5(
  apl: ApplicationLayer4 | ApplicationLayer12,
  ll: LinkLayer
) {
  const iv = Buffer.alloc(16, apl.accessNo);
  if (apl.ci === CI_RESP_4) {
    ll.addressRaw.copy(iv, 0);
  } else {
    iv.writeUInt16LE(apl.meterManufacturer, 0);
    iv.writeUInt32LE(apl.meterId, 2);
    iv.writeUInt8(apl.meterVersion, 6);
    iv.writeUInt8(apl.meterDevice, 7);
  }
  return iv;
}

function isPrios(linkLayer: LinkLayer | WiredLinkLayer, ci: number) {
  return (
    isLinkLayer(linkLayer) &&
    linkLayer.manufacturer === "DME" &&
    CI_RESP_PRIOS.includes(ci)
  );
}

function isTechem(linkLayer: LinkLayer | WiredLinkLayer, ci: number) {
  return (
    isLinkLayer(linkLayer) &&
    linkLayer.manufacturer === "TCH" &&
    CI_RESP_TECHEM.includes(ci)
  );
}
export async function decodeApplicationLayer(
  state: ParserState,
  linkLayer: LinkLayer | WiredLinkLayer,
  afl?: AuthenticationAndFragmentationLayer
): Promise<{
  state: ParserState;
  applicationLayer: ApplicationLayer;
  linkLayer: LinkLayer;
}> {
  const data = state.data;
  let pos = state.pos;

  const offset = pos;
  const ci = data[pos++];

  if (ci === CI_RESP_SML_4 || ci === CI_RESP_SML_12) {
    throw new Error("Payload is SML encoded. SML decoding is not implemented");
  }

  if (ci === CI_RESP_0) {
    return {
      state: { data: data, pos: pos },
      applicationLayer: { ci, offset },
      linkLayer: getDummyLinkLayer(linkLayer),
    };
  } else if (ci === CI_RESP_COMPACT) {
    const { newPos, apl } = getApplicationLayerCompact(data, pos, offset);
    return {
      state: { data: data, pos: newPos },
      applicationLayer: apl,
      linkLayer: getDummyLinkLayer(linkLayer),
    };
  }

  let apl: ApplicationLayer4 | ApplicationLayer12;
  let ll: LinkLayer;

  if (ci === CI_RESP_4 || ci === CI_RESP_12) {
    const res =
      ci === CI_RESP_4
        ? getApplicationLayer4(data, pos, offset)
        : getApplicationLayer12(data, pos, offset);
    pos = res.newPos;
    apl = res.apl;
    ll = mockLinkLayerFromApplicationLayer(apl, linkLayer);
  } else if (isPrios(linkLayer, ci)) {
    return await decodePriosApplicationLayer(data, offset, linkLayer);
  } else if (isTechem(linkLayer, ci)) {
    return await decodeTechemApplicationLayer(
      data,
      offset,
      linkLayer as LinkLayer
    );
  } else {
    throw new Error(
      `Unsupported CI Field 0x${ci.toString(16)}\nremaining payload is ${data.toString("hex", pos)}`
    );
  }

  let decryptedData: Buffer;

  if (apl.config.mode !== 0) {
    if (state.key === undefined) {
      throw new Error("Encrypted telegram but no AES key provided");
    }

    if (apl.config.mode === 5) {
      const encryptedLength = apl.config.encryptedBlocks * AES_BLOCK_SIZE;
      decryptedData = decryptInPlace(
        data,
        state.key,
        createIvMode5(apl, ll),
        "aes-128-cbc",
        pos,
        encryptedLength
      );
    } else if (apl.config.mode === 7) {
      if (afl === undefined) {
        throw new Error("Mode 7 encryption but no AFL found");
      }
      const encryptedLength = apl.config.encryptedBlocks * AES_BLOCK_SIZE;
      const { kenc, kmac } = await calcKenc(state.key, {
        ll,
        afl,
        apl,
      });
      const msgData = data.subarray(apl.offset, pos + encryptedLength);
      await checkAflMac(kmac, msgData, afl);

      decryptedData = decryptInPlace(
        data,
        kenc,
        createIvMode7(),
        "aes-128-cbc",
        pos,
        encryptedLength
      );
    } else {
      throw new Error(
        `Encryption mode ${apl.config.mode.toString(16)} not implemented`
      );
    }
  }

  return {
    state: {
      ...state,
      pos: pos,
      data: decryptedData ?? data,
    },
    applicationLayer: apl,
    linkLayer: ll,
  };
}
