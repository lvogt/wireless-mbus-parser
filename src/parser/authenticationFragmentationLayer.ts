import { CI_AFL } from "@/helper/constants";
import { ParserError } from "@/helper/error";
import { log } from "@/helper/logger";
import type { AuthenticationAndFragmentationLayer, ParserState } from "@/types";

export function hasAuthenticationAndFragmentationLayer(state: ParserState) {
  return state.data[state.pos] === CI_AFL;
}

export function decodeAuthenticationAndFragmentationLayer(state: ParserState): {
  state: ParserState;
  authenticationAndFragmentationLayer:
    | AuthenticationAndFragmentationLayer
    | undefined;
} {
  if (!hasAuthenticationAndFragmentationLayer(state)) {
    return { state, authenticationAndFragmentationLayer: undefined };
  }

  const data = state.data;
  let pos = state.pos;

  const ci = data[pos++];
  if (ci !== CI_AFL) {
    throw new ParserError("UNEXPECTED_STATE", `Wrong AFL CI: ${ci}`);
  }
  log.debug("Authentification and Fragmentation Layer");

  const afll = data[pos++];

  const fclRaw = data.readUInt16LE(pos);
  pos += 2;

  const fcl: AuthenticationAndFragmentationLayer["fcl"] = {
    /* 0b1000000000000000 - reserved */
    mf:
      (fclRaw & 0b0100000000000000) !==
      0 /* More fragments: 0 last fragment; 1 more following */,
    mclp:
      (fclRaw & 0b0010000000000000) !==
      0 /* Message Control Field present in fragment */,
    mlp:
      (fclRaw & 0b0001000000000000) !==
      0 /* Message Length Field present in fragment */,
    mcrp:
      (fclRaw & 0b0000100000000000) !==
      0 /* Message Counter Field present in fragment */,
    macp:
      (fclRaw & 0b0000010000000000) !== 0 /* MAC Field present in fragment */,
    kip:
      (fclRaw & 0b0000001000000000) !==
      0 /* Key Information present in fragment */,
    /* 0b0000000100000000 - reserved */
    fid: fclRaw & 0b0000000011111111 /* fragment ID */,
  };

  let mclRaw: number;
  let mcl: AuthenticationAndFragmentationLayer["mcl"];

  if (fcl.mclp) {
    // AFL Message Control Field (AFL.MCL)
    mclRaw = data[pos++];
    mcl = {
      /* 0b10000000 - reserved */
      mlmp:
        (mclRaw & 0b01000000) !==
        0 /* Message Length Field present in message */,
      mcmp:
        (mclRaw & 0b00100000) !==
        0 /* Message Counter Field present in message */,
      kimp:
        (mclRaw & 0b00010000) !==
        0 /* Key Information Field present in message */,
      at: mclRaw & 0b00001111 /* Authentication-Type */,
    };
  }

  let kiRaw: number;
  let ki: AuthenticationAndFragmentationLayer["ki"];

  if (fcl.kip) {
    // AFL Key Information Field (AFL.KI)
    kiRaw = data.readUInt16LE(pos);
    pos += 2;
    ki = {
      keyVersion: (kiRaw & 0b1111111100000000) >> 8,
      /* 0b0000000011000000 - reserved */
      kdfSelection: (kiRaw & 0b0000000000110000) >> 4,
      keyId: kiRaw & 0b0000000000001111,
    };
  }

  let mcr: number;

  if (fcl.mcrp) {
    // AFL Message Counter Field (AFL.MCR)
    mcr = data.readUInt32LE(pos);
    pos += 4;
  }

  let mac: Buffer;

  if (fcl.macp) {
    if (mcl === undefined) {
      throw new ParserError(
        "UNEXPECTED_STATE",
        "AFL MAC should be present but MCL is missing"
      );
    }
    // AFL MAC Field (AFL.MAC)
    // length of the MAC field depends on AFL.MCL.AT indicated by the AFL.MCL field
    // currently only AT = 5 is used (AES-CMAC-128 8bytes truncated)
    let mac_len = 0;
    if (mcl.at == 4) {
      mac_len = 4;
    } else if (mcl.at == 5) {
      mac_len = 8;
    } else if (mcl.at == 6) {
      mac_len = 12;
    } else if (mcl.at == 7) {
      mac_len = 16;
    }
    mac = Buffer.from(data.subarray(pos, pos + mac_len));
    pos += mac_len;
    log.debug(`AFL MAC ${mac.toString("hex")}`);
  }

  let ml: number;

  if (fcl.mlp) {
    // AFL Message Length Field (AFL.ML)
    ml = data.readUInt16LE(pos);
    pos += 2;
  }

  if (fcl.mf) {
    throw new ParserError(
      "UNIMPLEMENTED_FEATURE",
      "Fragmented messages are not supported yet."
    );
  }

  return {
    state: {
      ...state,
      pos,
    },
    authenticationAndFragmentationLayer: {
      ci,
      afll,
      fclRaw,
      fcl,
      mclRaw,
      mcl,
      kiRaw,
      ki,
      mcr,
      mac,
      ml,
    },
  };
}
