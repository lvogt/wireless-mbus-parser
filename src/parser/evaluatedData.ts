import { ParserError } from "@/helper/error";
import { isPrimaryVifString } from "@/helper/helper";
import { log } from "@/helper/logger";
import {
  applyFunctionFieldType,
  applyNumberOrStringifyDefault,
  applyStringifyDefault,
  extendDescription,
} from "@/helper/vifHelper";
import type {
  DataRecord,
  EvaluatedData,
  MeterData,
  PrimaryVif,
  VIFDescriptor,
  VIFEDescriptor,
} from "@/types";
import { VifTable } from "@/types";
import { defaultVIFs } from "@/vif/defaultVifs";
import { fbVifs } from "@/vif/fbVifs";
import { fdVifs } from "@/vif/fdVifs";
import {
  manufacturerSpecificsVifes,
  manufacturerSpecificsVifs,
} from "@/vif/manufacturerSpecificVifs";
import { vifExtensions } from "@/vif/vifExtension";

function getDescriptor(dataRecord: DataRecord, meterType: MeterData) {
  switch (dataRecord.header.vib.primary.table) {
    case VifTable.Default:
      return defaultVIFs.find(
        (item) => item.vif === dataRecord.header.vib.primary.vif
      );
    case VifTable.FD:
      return fdVifs.find(
        (item) => item.vif === dataRecord.header.vib.primary.vif
      );
    case VifTable.FB:
      return fbVifs.find(
        (item) => item.vif === dataRecord.header.vib.primary.vif
      );
    case VifTable.Plain:
      return getPlainTextDescriptor(dataRecord.header.vib.primary);
    case VifTable.Manufacturer:
      return getManufacturerSpecificsVifDescriptor(dataRecord, meterType);
    default:
      throw new ParserError(
        "UNIMPLEMENTED_FEATURE",
        "Table not yet implemented"
      );
  }
}

function getManufacturerSpecificsVifDescriptor(
  dataRecord: DataRecord,
  meterType: MeterData
) {
  if (meterType.manufacturer in manufacturerSpecificsVifs) {
    const descriptor = manufacturerSpecificsVifs[meterType.manufacturer].find(
      (item) => item.vif === dataRecord.header.vib.primary.vif
    );
    if (descriptor != undefined) {
      return descriptor;
    }
  }
  return getFallbackDescriptor(dataRecord, true);
}

function getManufacturerSpecificsVifeDescriptor(
  extension: number,
  meterType: MeterData
) {
  if (meterType.manufacturer in manufacturerSpecificsVifes) {
    const descriptor = manufacturerSpecificsVifes[meterType.manufacturer].find(
      (item) => item.vif === extension
    );
    if (descriptor != undefined) {
      return descriptor;
    }
  }
  return getFallbackExtensiontDescriptor(extension);
}

function getVifeDescriptor(
  extension: number,
  meterType: MeterData,
  manufacturerSpecific = false
) {
  if (manufacturerSpecific) {
    return getManufacturerSpecificsVifeDescriptor(extension, meterType);
  }

  const descriptor = vifExtensions.find((item) => item.vif === extension);
  if (descriptor === undefined) {
    return getFallbackExtensiontDescriptor(extension);
  } else {
    return descriptor;
  }
}

function getPlainTextDescriptor(primaryVif: PrimaryVif): VIFDescriptor {
  if (!isPrimaryVifString(primaryVif)) {
    throw new ParserError("UNEXPECTED_STATE", "PrimaryVifString expected!");
  }

  return {
    vif: primaryVif.vif,
    legacyName: "VIF_PLAIN_TEXT",
    unit: primaryVif.plainText,
    description: "",
    calc: (val) => val,
    apply: applyNumberOrStringifyDefault,
  };
}

function getFallbackDescriptor(
  dataRecord: DataRecord,
  manufacturerSpecific = false
): VIFDescriptor {
  return {
    vif: dataRecord.header.vib.primary.vif,
    legacyName: "VIF_UNKNOWN",
    unit: "",
    description: `Unknown ${manufacturerSpecific ? "manufacturer specific " : ""}VIF 0x${dataRecord.header.vib.primary.vif.toString(16)}`,
    calc: (val) => val,
    apply: applyStringifyDefault,
  };
}

function getFallbackExtensiontDescriptor(extension: number): VIFEDescriptor {
  return {
    vif: extension,
    legacyName: "VIFE_UNKNOWN",
    description: `Unknown VIFE 0x${extension.toString(16)}`,
    apply: extendDescription,
  };
}

function evaluatePrimaryVif(dataRecord: DataRecord, meterType: MeterData) {
  let descriptor = getDescriptor(dataRecord, meterType);
  if (descriptor === undefined) {
    descriptor = getFallbackDescriptor(dataRecord);
  }

  try {
    const evaluatedData = descriptor.apply(descriptor, dataRecord);
    return applyFunctionFieldType(evaluatedData, dataRecord);
  } catch {
    return applyStringifyDefault(descriptor, dataRecord);
  }
}

function evaluateVifExtension(
  data: EvaluatedData,
  dataRecord: DataRecord,
  meterType: MeterData,
  extension: number,
  manufacturerSpecific: boolean
) {
  const descriptor = getVifeDescriptor(
    extension,
    meterType,
    manufacturerSpecific
  );

  try {
    return descriptor.apply(descriptor, dataRecord, data);
  } catch (e: unknown) {
    log.debug(`Applying VIFE failed: ${JSON.stringify(e)}`);
  }
}

function evaluateDataRecord(
  dataRecord: DataRecord,
  meterType: MeterData
): EvaluatedData {
  const evaluatedData = evaluatePrimaryVif(dataRecord, meterType);
  const manufacturerSpecificPrimaryVif =
    dataRecord.header.vib.primary.table === VifTable.Manufacturer;

  let lastExtManufacturerSpecific = false;
  for (const ext of dataRecord.header.vib.extensions) {
    if (ext === 0x7f) {
      lastExtManufacturerSpecific = true;
      continue;
    }

    const manufacturerSpecificTable =
      manufacturerSpecificPrimaryVif || lastExtManufacturerSpecific;

    evaluateVifExtension(
      evaluatedData,
      dataRecord,
      meterType,
      ext,
      manufacturerSpecificTable
    );
  }

  return evaluatedData;
}

export function evaluateDataRecords(
  dataRecords: DataRecord[],
  meterType: MeterData
): EvaluatedData[] {
  return dataRecords.map((record) => evaluateDataRecord(record, meterType));
}
