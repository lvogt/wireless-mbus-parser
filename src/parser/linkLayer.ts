import {
  decodeManufacturer,
  getDeviceType,
  getMeterId,
  isWiredMbusFrame,
} from "@/helper/helper";
import type { LinkLayer, ParserState, WiredLinkLayer } from "@/types";

export function decodeLinkLayer(state: ParserState): {
  state: ParserState;
  linkLayer: LinkLayer | WiredLinkLayer;
} {
  const { data, pos } = state;
  const isWired = isWiredMbusFrame(data);

  if (isWired) {
    state.pos += 6;
    const linkLayer: WiredLinkLayer = {
      lField: data[pos + 1],
      cField: data[pos + 4],
      aField: data[pos + 5],
    };
    return { state, linkLayer };
  }

  state.pos += 10;

  const linkLayer: LinkLayer = {
    lField: data[pos + 0],
    cField: data[pos + 1],
    mField: data.readUInt16LE(pos + 2),
    aField: data.readUInt32LE(pos + 4),
    version: data[pos + 8],
    type: data[pos + 9],

    addressRaw: Buffer.from(data.subarray(pos + 2, pos + 10)),
    aFieldRaw: Buffer.from(data.subarray(pos + 4, pos + 10)),

    manufacturer: decodeManufacturer(data.readUInt16LE(pos + 2)),
    typeString: getDeviceType(data[pos + 9]),
    meterId: getMeterId(data, pos + 4),
  };

  return { state, linkLayer };
}
