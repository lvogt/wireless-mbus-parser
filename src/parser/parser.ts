import { stripAnyCrc } from "@/crc/crcHandler";
import { ParserError } from "@/helper/error";
import { getMeterData, isCompactFrame } from "@/helper/helper";
import { decodeApplicationLayer } from "@/parser/applicationLayer";
import { decodeAuthenticationAndFragmentationLayer } from "@/parser/authenticationFragmentationLayer";
import {
  calcHeaderCrc,
  decodeDataRecords,
  extractDataRecordHeaders,
} from "@/parser/dataRecords";
import { evaluateDataRecords } from "@/parser/evaluatedData";
import { decodeExtendedLinkLayer } from "@/parser/extendedLinkLayer";
import { decodeLinkLayer } from "@/parser/linkLayer";
import type {
  ApplicationLayer,
  DataRecordHeader,
  ParserOptions,
  ParserOptionsCommon,
  ParserOptionsFull,
  ParserOptionsSimple,
  ParserResult,
  ParserResultVerbose,
  ParserState,
} from "@/types";

export class WirelessMbusParser {
  private dataRecordHeaderCache: Record<number, DataRecordHeader[] | null> = {};

  async parse(
    data: Buffer,
    options: ParserOptionsFull
  ): Promise<ParserResultVerbose>;
  async parse(
    data: Buffer,
    options?: ParserOptionsSimple
  ): Promise<ParserResult>;

  async parse(
    data: Buffer,
    options?: ParserOptions
  ): Promise<ParserResultVerbose | ParserResult> {
    if (options?.verbose) {
      return this.parseFullResult(data, options);
    } else {
      return this.parseSimple(data, options);
    }
  }

  private async parseFullResult(
    data: Buffer,
    options?: Partial<ParserOptionsCommon>
  ): Promise<ParserResultVerbose> {
    const crcFreeData = stripAnyCrc(data, options?.containsCrc);

    const state = {
      data: crcFreeData,
      pos: 0,
      key: options?.key ?? undefined,
    };

    const { state: llState, linkLayer: llFromLinkLayer } =
      decodeLinkLayer(state);

    const { state: ellState, extendedLinkLayer } = decodeExtendedLinkLayer(
      llState,
      llFromLinkLayer
    );

    const { state: aflState, authenticationAndFragmentationLayer } =
      decodeAuthenticationAndFragmentationLayer(ellState);

    const {
      state: aplState,
      applicationLayer,
      linkLayer,
    } = await decodeApplicationLayer(
      aflState,
      llFromLinkLayer,
      authenticationAndFragmentationLayer
    );
    const meterData = getMeterData(linkLayer, applicationLayer);

    const dataRecords = this.handleDataRecordDecoding(
      aplState,
      applicationLayer
    );

    const evaluatedData = evaluateDataRecords(dataRecords, meterData);

    return {
      data: evaluatedData,
      meter: meterData,
      linkLayer,
      extendedLinkLayer,
      authenticationAndFragmentationLayer,
      applicationLayer,
      dataRecords,
      rawData: aplState.data,
    };
  }

  private async parseSimple(
    data: Buffer,
    options?: Partial<ParserOptionsCommon>
  ): Promise<ParserResult> {
    const { data: evaluatedData, meter } = await this.parseFullResult(
      data,
      options
    );
    return { data: evaluatedData, meter };
  }

  private handleDataRecordDecoding(
    state: ParserState,
    applicationLayer: ApplicationLayer
  ) {
    if (isCompactFrame(applicationLayer)) {
      const dataRecordHeaders =
        this.dataRecordHeaderCache[applicationLayer.headerCrc];
      if (dataRecordHeaders === undefined || dataRecordHeaders === null) {
        this.dataRecordHeaderCache[applicationLayer.headerCrc] = null;
        throw new ParserError(
          "DATA_RECORD_CACHE_MISSING",
          "Compact frame received but data record cache is missing"
        );
      }
      const { dataRecords } = decodeDataRecords(state, dataRecordHeaders);
      return dataRecords;
    } else {
      const { dataRecords } = decodeDataRecords(state);

      const headerCrc = calcHeaderCrc(dataRecords, state.data);
      if (this.dataRecordHeaderCache[headerCrc] === null) {
        this.dataRecordHeaderCache[headerCrc] =
          extractDataRecordHeaders(dataRecords);
      }

      return dataRecords;
    }
  }
}
