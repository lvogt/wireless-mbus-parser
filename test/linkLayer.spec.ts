import { describe, expect, it } from "vitest";

import { decodeLinkLayer } from "@/parser/linkLayer";

describe("Link Layer", () => {
  it("Check fields", () => {
    const result = decodeLinkLayer({
      data: Buffer.from("2E449315785634003303", "hex"),
      pos: 0,
    });

    expect(result.state.pos).toBe(10);
    expect(result.isWired).toBe(false);
    expect(result.linkLayer).matchSnapshot();
  });

  it("Check fields with offset", () => {
    const result = decodeLinkLayer({
      data: Buffer.from("123456789a2E449315785634003377", "hex"),
      pos: 5,
    });

    expect(result.state.pos).toBe(15);
    expect(result.isWired).toBe(false);
    expect(result.linkLayer).matchSnapshot();
  });

  it("Check fields - Wired M-Bus", () => {
    const result = decodeLinkLayer({
      data: Buffer.from("68585868080516", "hex"),
      pos: 0,
    });

    expect(result.state.pos).toBe(6);
    expect(result.isWired).toBe(true);
    expect(result.linkLayer).matchSnapshot();
  });

  /*it("Check fields - Wired M-Bus - after Application Layer", () => {
    const parser = createParser(
      "6858586808057281052661A51170073E3000000C13000200008C1013000000008C2013000200003B3BDDBDEB0B26504701025ADB000266B900046D0C0B752B4C1300020000CC101300000000CC201300020000426C5F2C42EC7E7F2CDD16",
      0
    );
    parser.decodeLinkLayer();
    parser.decodeApplicationLayer();

    expect(parser.telegram.linkLayer).to.eql({
      lField: 0x58,
      cField: 0x08,
      mField: 0x11a5,
      aField: 0x05,
      version: 0x70,
      type: 0x07,
      addressRaw: Buffer.from("A511810526617000", "hex"),
      aFieldRaw: Buffer.from("810526617000", "hex"),
      manufacturer: "DME",
      typeString: "Water",
      meterId: "02001312",
    });
  }); */
});
