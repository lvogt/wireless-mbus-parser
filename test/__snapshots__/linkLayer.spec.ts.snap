// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Link Layer > Check fields - Wired M-Bus 1`] = `
{
  "aField": 5,
  "cField": 8,
  "lField": 88,
}
`;

exports[`Link Layer > Check fields 1`] = `
{
  "aField": 3430008,
  "aFieldRaw": {
    "data": [
      120,
      86,
      52,
      0,
      51,
      3,
    ],
    "type": "Buffer",
  },
  "addressRaw": {
    "data": [
      147,
      21,
      120,
      86,
      52,
      0,
      51,
      3,
    ],
    "type": "Buffer",
  },
  "cField": 68,
  "lField": 46,
  "mField": 5523,
  "manufacturer": "ELS",
  "meterId": "00345678",
  "type": 3,
  "typeString": "Gas",
  "version": 51,
}
`;

exports[`Link Layer > Check fields with offset 1`] = `
{
  "aField": 3430008,
  "aFieldRaw": {
    "data": [
      120,
      86,
      52,
      0,
      51,
      119,
    ],
    "type": "Buffer",
  },
  "addressRaw": {
    "data": [
      147,
      21,
      120,
      86,
      52,
      0,
      51,
      119,
    ],
    "type": "Buffer",
  },
  "cField": 68,
  "lField": 46,
  "mField": 5523,
  "manufacturer": "ELS",
  "meterId": "00345678",
  "type": 119,
  "typeString": "unknown",
  "version": 51,
}
`;
